### PR TITLE
feat(card): refactor card component

### PR DESCRIPTION
**Pull request by helen-v**
## Screenshots:
**Before** (If applicable):

**After**:

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [x] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Other... Please describe:

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: web-components/src/components/button/scss/settings.scss [hunk 1]
```diff
@@ -222,7 +222,7 @@ $button--duck-egg__border: 0 0 0 1px $md-gray-50 !default;
 $button--violet__background-color: $md-violet-50 !default;
 $button--violet__background-color--hover: $md-violet-60 !default;
 $button--violet__background-color--active: $md-violet-70 !default;
-$button--violet__border: 0 0 0 1px $md-violet-90 !default;
```
Comment (futurethang) [inline web-components/src/components/button/scss/settings.scss:225]: @helen-v Is this card related or responding to another issue?

File: web-components/src/components/favorite/Favorite.ts [hunk 2]
```diff
@@ -88,18 +87,20 @@ export namespace Favorite {
...
             value=${this.value}
             ?checked=${this.checked}
             ?disabled=${this.disabled}
             aria-hidden="true"
             name="favorite-checkbox"
-            />
-            <md-icon name="${this.getIconName()}" color="${this.checked ? "yellow" : nothing}"></md-icon>
+          />
+          <md-icon name="${this.getIconName()}" color="${this.checked ? "yellow" : nothing}"></md-icon>
         </label>
```
Comment (futurethang) [inline web-components/src/components/favorite/Favorite.ts:104]: Root cause is unknown, but I cannot toggle the Favorite star by using Space or Enter key. I can focus with Tab just fine, but cannot change its state.

File: web-components/src/components/card/Card.ts [hunk 3]
```diff
@@ -1,53 +1,153 @@
...
+    get cardClassMap() {
+      return {
+        "full-screen": this.full
+      };
+    }
+
+    connectedCallback() {
+      super.connectedCallback();
+    }
+
```
Comment (futurethang) [inline web-components/src/components/card/Card.ts:47]: this can be removed, it's only needed if we are adding more connectedCallback operations

**Review by futurethang — CHANGES_REQUESTED** [review]

File: web-components/src/components/card/Card.ts [hunk 4]
```diff
@@ -1,53 +1,153 @@
...
     @property({ type: String }) title = "title";
     @property({ type: String }) subtitle = "";
-    @property({ type: String }) src = "";
-    @property({ type: String }) link = "";
-    @property({ type: String }) linkTitle = "";
-    @property({ type: String }) buttonTitle = "";
+    @property({ type: String }) id = "";
+    @property({ type: Boolean }) fullscreen = false;
+    @property({ type: String }) info = "";
+    @property({ type: Array }) menuOption: (string | { [key: string]: string })[] = [];
```
Comment (yananym) [inline web-components/src/components/card/Card.ts:30]: menuOptions - since it is an array it has to be plural

File: web-components/src/components/card/Card.ts [hunk 5]
```diff
@@ -1,53 +1,153 @@
...
   @customElementWithCheck("md-card")
   export class ELEMENT extends LitElement {
     @property({ type: String }) title = "title";
     @property({ type: String }) subtitle = "";
-    @property({ type: String }) src = "";
-    @property({ type: String }) link = "";
-    @property({ type: String }) linkTitle = "";
-    @property({ type: String }) buttonTitle = "";
+    @property({ type: String }) id = "";
+    @property({ type: Boolean }) fullscreen = false;
```
Comment (yananym) [inline web-components/src/components/card/Card.ts:28]: I would omit this for now. I told Mocha earlier - having card expand defeats the purpose of having a card.

File: web-components/src/[sandbox]/examples/card.ts [hunk 6]
```diff
@@ -1,11 +1,35 @@
...
 export const cardTemplate = html`
   <h3>Default Card</h3>
-  <md-card>
+  <md-card
+    .menuOption=${cardMenu}
+    id="123456789"
+    title="Team A Report - Q1"
+    subtitle="Updated 2 hours ago"
+    fullscreen
+    @card-click=${(e: MouseEvent) => {
```
Comment (yananym) [inline web-components/src/[sandbox]/examples/card.ts:14]: can you show the even handling here? e.g. how to handle "Edit" click vs other menu items being clicked
Also the card itself needs to be clickable - and the menu items should emit a different custom event when clicked

File: web-components/src/components/card/Card.ts [hunk 7]
```diff
@@ -1,53 +1,153 @@
...
+
+    handleToggleExpandCollapse() {
+      this.full = !this.full;
+    }
+
+    handleCardMenuEvent(event: MouseEvent, label: string) {
+      event.preventDefault();
+      const ev = label.toLowerCase();
+      this.dispatchEvent(
+        new CustomEvent<{ id: string; type: string }>("card-click", {
```
Comment (yananym) [inline web-components/src/components/card/Card.ts:56]: i would rename it to card-menu-click. Card needs to be clickable by itself

File: web-components/src/components/card/Card.ts [hunk 8]
```diff
@@ -1,53 +1,153 @@
...
-              <div class="md-card-header-actions">
-                <md-menu-overlay placement="bottom-end">
-                  <md-button variant="white" slot="menu-trigger" circle>
-                    <md-icon slot="icon" name="icon-more-adr_16"></md-icon>
-                  </md-button>
-                  <md-list></md-list>
-                </md-menu-overlay>
-              </div>
+            <slot name="card-header-icon">
+              <md-favorite value="${`card-` + this.id}"></md-favorite>
```
Comment (yananym) [inline web-components/src/components/card/Card.ts:85]: it is not really an icon slot. it can be anything: an icon, a favorite button (button is not just the icon. it is an action) an avatar. an image...
what can we name it to be generic enough?.. `card-header-aside`

File: web-components/src/components/card/Card.ts [hunk 9]
```diff
@@ -1,53 +1,153 @@
...
+                      ariaLabel="Maximaze Card"
+                      class="md-card-max-icon"
+                      color="color-none"
+                      size="size-none"
+                    >
+                      <md-icon slot="icon" name=${this.full ? "minimize_16" : "maximize_16"}></md-icon>
+                    </md-button>
+                  `
+                : nothing}
+              <md-menu-overlay class="md-card-menu" placement="bottom-start">
```
Comment (yananym) [inline web-components/src/components/card/Card.ts:116]: we should only show it if `this.menuOption` is not empty

File: web-components/src/components/favorite/Favorite.ts [hunk 10]
```diff
@@ -39,11 +39,10 @@ export namespace Favorite {
     }
 
     getIconName() {
-      switch (this.checked) {
-        case true:
-          return "favorite-active_16";
-        default:
-          return "favorite_16";
+      if (this.checked) {
```
Comment (yananym) [inline web-components/src/components/favorite/Favorite.ts:42]: use a shorthand: `return this.checked ? "favorite-active_16" :  "favorite_16";`

**Comment by yananym** [comment]
Card-image.png file needs to only be there for sandbox. I would simply reference an external image instead and not add any assets to the project

File: web-components/src/components/button/scss/settings.scss (hunk 1 above)
Comment (helen-v) [reply web-components/src/components/button/scss/settings.scss:225]: Only here noticed that the outline border for violet does not match .

File: web-components/src/components/favorite/Favorite.ts (hunk 10 above)
Comment (helen-v) [reply web-components/src/components/favorite/Favorite.ts:42]: Done

File: web-components/src/components/card/Card.ts (hunk 9 above)
Comment (helen-v) [reply web-components/src/components/card/Card.ts:116]: Done

File: web-components/src/components/favorite/Favorite.ts (hunk 2 above)
Comment (helen-v) [reply web-components/src/components/favorite/Favorite.ts:104]: Fixed

File: web-components/src/components/card/Card.ts (hunk 3 above)
Comment (helen-v) [reply web-components/src/components/card/Card.ts:47]: Agree. Removed.

File: web-components/src/components/card/Card.ts (hunk 4 above)
Comment (helen-v) [reply web-components/src/components/card/Card.ts:30]: Done

File: web-components/src/components/card/Card.ts (hunk 8 above)
Comment (helen-v) [reply web-components/src/components/card/Card.ts:85]: Renamed

**Comment by helen-v** [comment]
> Card-image.png file needs to only be there for sandbox. I would simply reference an external image instead and not add any assets to the project

Done.

File: web-components/src/[sandbox]/examples/card.ts (hunk 6 above)
Comment (helen-v) [reply web-components/src/[sandbox]/examples/card.ts:14]: Done

File: web-components/src/components/card/Card.ts (hunk 5 above)
Comment (helen-v) [reply web-components/src/components/card/Card.ts:28]: Ok

File: web-components/src/components/card/Card.ts [hunk 11]
```diff
@@ -1,53 +1,189 @@
...
-import { html, LitElement, property } from "lit-element";
+import { html, internalProperty, LitElement, property } from "lit-element";
 import styles from "./scss/module.scss";
 import "@/components/icon/Icon";
 import "@/components/button/Button";
-import { ifDefined } from "lit-html/directives/if-defined";
+import "@/components/list/List";
+import "@/components/list/ListItem";
+import "@/components/favorite/Favorite";
+import { Key } from "@/constants";
```
Comment (yananym) [inline web-components/src/components/card/Card.ts:18]: please use relative path for type

Comment (yananym) [reply web-components/src/components/card/Card.ts:18]: otherwise it breaks the Angular build

File: web-components/src/components/card/Card.ts (hunk 5 above)
Comment (yananym) [reply web-components/src/components/card/Card.ts:29]: just remove everything that has to do with full screen. until further notice

File: web-components/src/components/card/Card.ts [hunk 12]
```diff
@@ -1,53 +1,189 @@
...
       return [reset, styles];
     }
 
+    get cardClassMap() {
+      return {
+        "full-screen": this.full
+      };
+    }
+
+    handleToggleExpandCollapse(event: MouseEvent) {
```
Comment (yananym) [inline web-components/src/components/card/Card.ts:45]: remove this

File: web-components/src/components/card/Card.ts [hunk 13]
```diff
@@ -1,53 +1,189 @@
...
-    @property({ type: String }) src = "";
-    @property({ type: String }) link = "";
-    @property({ type: String }) linkTitle = "";
-    @property({ type: String }) buttonTitle = "";
+    @property({ type: String }) id = "";
+    @property({ type: Boolean }) fullscreen = false;
+    @property({ type: String }) info = "";
+    @property({ type: Array }) menuOptions: (string | any)[] = [];
+
+    @internalProperty() private full = false;
```
Comment (yananym) [inline web-components/src/components/card/Card.ts:33]: remove this

File: web-components/src/components/card/Card.ts [hunk 14]
```diff
@@ -1,53 +1,189 @@
...
 
     static get styles() {
       return [reset, styles];
     }
 
+    get cardClassMap() {
```
Comment (yananym) [inline web-components/src/components/card/Card.ts:39]: remove this

File: web-components/src/components/card/Card.ts [hunk 15]
```diff
@@ -1,53 +1,189 @@
...
+              ${this.info
+                ? html`
+                    <md-tooltip message="${this.info}" placement="bottom">
+                      <md-button ariaLabel="Card Info" class="md-card-info-icon" color="color-none" size="size-none">
+                        <md-icon slot="icon" name="info_16"></md-icon>
+                      </md-button>
+                    </md-tooltip>
+                  `
+                : nothing}
+              ${this.fullscreen
```
Comment (yananym) [inline web-components/src/components/card/Card.ts:135]: remove this

File: web-components/src/components/card/Card.ts [hunk 16]
```diff
@@ -1,53 +1,189 @@
...
+              <slot name="card-header-title">
+                <h1>${this.title}</h1>
+                <h2>${this.subtitle}</h2>
+              </slot>
+            </div>
+            <div class="md-card-header-actions">
+              ${this.info
+                ? html`
+                    <md-tooltip message="${this.info}" placement="bottom">
+                      <md-button ariaLabel="Card Info" class="md-card-info-icon" color="color-none" size="size-none">
```
Comment (yananym) [inline web-components/src/components/card/Card.ts:129]: aria label needs to be localized through prop

File: web-components/src/components/card/Card.ts [hunk 17]
```diff
@@ -1,53 +1,189 @@
...
+                    >
+                      <md-icon slot="icon" name=${this.full ? "minimize_16" : "maximize_16"}></md-icon>
+                    </md-button>
+                  `
+                : nothing}
+              ${this.menuOptions.length != 0
+                ? html`
+                    <md-menu-overlay class="md-card-menu" placement="bottom-start">
+                      <md-button
+                        ariaLabel="Menu"
```
Comment (yananym) [inline web-components/src/components/card/Card.ts:152]: arial label needs to be localized through prop

File: web-components/src/components/card/Card.ts (hunk 5 above)
Comment (helen-v) [reply web-components/src/components/card/Card.ts:29]: Remove fullscreen mode

File: web-components/src/components/card/Card.ts (hunk 11 above)
Comment (helen-v) [reply web-components/src/components/card/Card.ts:18]: Done

File: web-components/src/components/card/Card.ts (hunk 12 above)
Comment (helen-v) [reply web-components/src/components/card/Card.ts:45]: Done

File: web-components/src/components/card/Card.ts (hunk 13 above)
Comment (helen-v) [reply web-components/src/components/card/Card.ts:33]: Done

File: web-components/src/components/card/Card.ts (hunk 14 above)
Comment (helen-v) [reply web-components/src/components/card/Card.ts:39]: Done

File: web-components/src/components/card/Card.ts (hunk 16 above)
Comment (helen-v) [reply web-components/src/components/card/Card.ts:129]: Done

File: web-components/src/components/card/Card.ts (hunk 17 above)
Comment (helen-v) [reply web-components/src/components/card/Card.ts:152]: Done

File: web-components/src/components/card/Card.ts [hunk 18]
```diff
@@ -1,53 +1,168 @@
...
+
 import reset from "@/wc_scss/reset.scss";
-import { customElementWithCheck } from "@/mixins/CustomElementCheck";
+import { customElementWithCheck } from "../../mixins/CustomElementCheck";
 import { html, LitElement, property } from "lit-element";
 import styles from "./scss/module.scss";
-import "@/components/icon/Icon";
-import "@/components/button/Button";
-import { ifDefined } from "lit-html/directives/if-defined";
+import "../icon/Icon";
```
Comment (yananym) [inline web-components/src/components/card/Card.ts:13]:
```suggestion
import "@/components/icon/Icon";
```

File: web-components/src/components/card/Card.ts [hunk 19]
```diff
@@ -1,53 +1,168 @@
...
+import "../button/Button";
```
Comment (yananym) [inline web-components/src/components/card/Card.ts:14]:
```suggestion
import "@/components/button/Button";
```

File: web-components/src/components/card/Card.ts [hunk 20]
```diff
@@ -1,53 +1,168 @@
...
+import "../list/List";
```
Comment (yananym) [inline web-components/src/components/card/Card.ts:15]:
```suggestion
import "@/components/list/List";
```

File: web-components/src/components/card/Card.ts [hunk 21]
```diff
@@ -1,53 +1,168 @@
...
+import "../list/ListItem";
```
Comment (yananym) [inline web-components/src/components/card/Card.ts:16]:
```suggestion
import "@/components/list/ListItem";
```

File: web-components/src/components/card/Card.ts [hunk 22]
```diff
@@ -1,53 +1,168 @@
...
+import "../favorite/Favorite";
```
Comment (yananym) [inline web-components/src/components/card/Card.ts:17]:
```suggestion
import "@/components/favorite/Favorite";
```

File: web-components/src/components/card/Card.ts [hunk 23]
```diff
@@ -1,53 +1,168 @@
...
+import { Key } from "../../constants";
```
Comment (yananym) [inline web-components/src/components/card/Card.ts:18]:
```suggestion
import { Key } from "../../constants"; // Keep type import as a relative path
```

File: web-components/src/components/card/Card.ts [hunk 24]
```diff
@@ -1,53 +1,168 @@
...
+ * Copyright (c) Cisco Systems, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ */
+
 import reset from "@/wc_scss/reset.scss";
-import { customElementWithCheck } from "@/mixins/CustomElementCheck";
+import { customElementWithCheck } from "../../mixins/CustomElementCheck";
```
Comment (yananym) [inline web-components/src/components/card/Card.ts:10]:
```suggestion
import { customElementWithCheck } from "@/mixins/CustomElementCheck";
```

**Review by yananym — APPROVED** [review]
